File: aoc_2022/src/main/java/me/frenz/day13/Day13.java
package me.frenz.day13;

import me.frenz.Day;
import me.frenz.Pair;

import java.util.ArrayList;
import java.util.List;


public class Day13 extends Day<Integer, Integer> {

    private final PacketParser packetParser = new PacketParser();

    public Day13(final List<String> input) {
        super(input.stream().filter(it -> !it.isEmpty()).toList());
    }

    @Override
    protected Integer part1() {
        final List<Pair<PacketValue>> parseInput = parseInput();
        final List<Integer> sum = new ArrayList<>(parseInput.size());
        for (int i = 0; i < parseInput.size(); i++) {
            final Pair<PacketValue> packets = parseInput.get(i);
            final PacketValue left = packets.left(), right = packets.right();

            if (compare(left, right) > 0) {
                sum.add(i + 1);
            }
        }

        return sum.stream().reduce(Integer::sum).orElse(-1);
    }

    @Override
    protected Integer part2() {
        return -1;
    }

    List<Pair<PacketValue>> parseInput() {
        assert input.size() % 2 == 0 : "Input lines not divisible by two! Got %s lines".formatted(input.size());
        final List<Pair<PacketValue>> packetPairs = new ArrayList<>();

        for (int i = 0; i < input.size(); i += 2) {
            final ContainerPacket left = new ContainerPacket();
            final ContainerPacket right = new ContainerPacket();
            packetParser.parsePacket(left, input.get(i));
            packetParser.parsePacket(right, input.get(i + 1));

            packetPairs.add(new Pair<>(left.unwrap(), right.unwrap()));
        }

        return packetPairs;
    }

    int compare(PacketValue left, PacketValue right) {
        if (left instanceof IntPacket li && right instanceof IntPacket ri) {
            return Integer.compare(ri.get(), li.get());
        } else if (left instanceof IntPacket li && right instanceof ContainerPacket rc) {
            return compare(new ContainerPacket(List.of(li)), rc);
        } else if (left instanceof ContainerPacket lc && right instanceof IntPacket ri) {
            return compare(lc, new ContainerPacket(List.of(ri)));
        } else if (left instanceof ContainerPacket lc && right instanceof ContainerPacket rc) {
            int areEqual = 0; // 0 means continue checking
            final int leftSize = lc.get().size();
            final int rightSize = rc.get().size();
            for (int i = 0; i < leftSize; i++) {
                if (i >= rightSize) {
                    areEqual = -1;
                    break;
                }
                areEqual = compare(lc.get().get(i), rc.get().get(i));
            }
            return rightSize > leftSize ? 1 : areEqual;
        } else {
            throw new IllegalArgumentException("This should never happen");
        }
    }
}

File: aoc_2022/src/main/java/me/frenz/day13/Day13.kt
package me.frenz.day13

import me.frenz.Day
import me.frenz.Pair
import java.util.*
import kotlin.collections.ArrayList

class Day13(input: List<String>) : Day<Int?, Int?>(input.filter(String::isNotBlank)) {
    private val packetParser = PacketParser()

    val parsedInput: List<Pair<PacketValue>>
        get() = parseInput()

    public override fun part1(): Int {
        val parsedInput = parsedInput
        val sum: MutableList<Int> = ArrayList(parsedInput.size)
        for (i in parsedInput.indices) {
            val packets = parsedInput[i]
            if (compare(packets.left, packets.right) > 0) {
                sum.add(i + 1)
            }
        }
        return Optional.ofNullable(sum.reduceOrNull(Integer::sum)).orElse(0)
    }

    public override fun part2(): Int {
        return -1
    }

    private fun parseInput(): List<Pair<PacketValue>> {
        assert(input.size % 2 == 0) { "Input lines not divisible by two! Got %s lines".format(input.size) }
        return input.windowed(2, step = 2)
            .map { Pair(packetParser.parsePacket(it[0]), packetParser.parsePacket(it[1])) }
            .toList()
    }

    fun compare(left: PacketValue, right: PacketValue): Int {
        return when {
            left is IntPacket && right is IntPacket ->
                right.get().compareTo(left.get())
            left is IntPacket && right is ContainerPacket ->
                compare(ContainerPacket(listOf(left)), right)
            left is ContainerPacket && right is IntPacket ->
                compare(left, ContainerPacket(listOf(right)))
            left is ContainerPacket && right is ContainerPacket ->
                compareContainerPackets(left, right)
            else -> throw IllegalArgumentException("This should never happen")
        }
    }

    private fun compareContainerPackets(left: ContainerPacket, right: ContainerPacket): Int {
        val lPackets = left.get().size
        val rPackets = right.get().size
        var idx = 0
        while (idx < lPackets && idx < rPackets) {
            when (compare(left.get()[idx]!!, right.get()[idx]!!)) {
                1 -> return 1
                -1 -> return -1
                else -> idx++
            }
        }
        return when {
            idx == lPackets && idx != rPackets -> 1
            idx == rPackets && idx != lPackets -> -1
            else -> 0
        }
    }
}
